// @joplin/lib/services/interop/InteropService_Exporter_Html.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const InteropService_1 = require("../../services/interop/InteropService");
const test_utils_1 = require("../../testing/test-utils");
const Folder_1 = require("../../models/Folder");
const Note_1 = require("../../models/Note");
const fs = require("fs-extra");
const test_utils_2 = require("../../testing/test-utils");
const types_1 = require("../../services/plugins/api/types");
const types_2 = require("./types");
function recreateExportDir() {
    return __awaiter(this, void 0, void 0, function* () {
        const dir = (0, test_utils_1.exportDir)();
        yield fs.remove(dir);
        yield fs.mkdirp(dir);
    });
}
describe('interop/InteropService_Exporter_Html', function () {
    beforeEach((done) => __awaiter(this, void 0, void 0, function* () {
        yield (0, test_utils_1.setupDatabaseAndSynchronizer)(1);
        yield (0, test_utils_1.switchClient)(1);
        yield recreateExportDir();
        done();
    }));
    test('should export HTML file', (() => __awaiter(this, void 0, void 0, function* () {
        const service = InteropService_1.default.instance();
        const folder1 = yield Folder_1.default.save({ title: 'folder1' });
        yield Note_1.default.save({ body: '**ma note**', parent_id: folder1.id });
        const filePath = `${(0, test_utils_1.exportDir)()}/test.html`;
        yield service.export({
            path: filePath,
            format: 'html',
            packIntoSingleFile: false,
        });
        const content = yield fs.readFile(filePath, 'utf8');
        expect(content).toContain('<strong>ma note</strong>');
    })));
    test('should export HTML directory', (() => __awaiter(this, void 0, void 0, function* () {
        const service = InteropService_1.default.instance();
        const folder1 = yield Folder_1.default.save({ title: 'folder1' });
        yield Folder_1.default.save({ title: 'folder2' });
        yield Note_1.default.save({ title: 'note1', parent_id: folder1.id });
        yield Note_1.default.save({ title: 'note2', parent_id: folder1.id });
        const dir = (0, test_utils_1.exportDir)();
        yield service.export({
            path: dir,
            format: 'html',
            target: types_2.FileSystemItem.Directory,
        });
        const rootDirs = yield fs.readdir(dir);
        rootDirs.sort();
        expect(rootDirs).toEqual(['folder1', 'folder2']);
        const files = yield fs.readdir(`${dir}/${rootDirs[0]}`);
        expect(files).toContain('note1.html');
        expect(files).toContain('note2.html');
    })));
    test('should export plugin assets', (() => __awaiter(this, void 0, void 0, function* () {
        const service = InteropService_1.default.instance();
        const folder1 = yield Folder_1.default.save({ title: 'folder1' });
        yield Note_1.default.save({ body: '**ma note**', parent_id: folder1.id });
        const filePath = `${(0, test_utils_1.exportDir)()}/test.html`;
        const contentScriptPath = (0, test_utils_2.tempFilePath)('js');
        yield fs.writeFile(contentScriptPath, `module.exports = {
			default: function(_context) { 
				return {
					plugin: function (markdownIt, _options) {
						
					}, 
					assets: function() {
						return [
							{ name: 'fence.css' }
						];
					},		
				}
			},
		}`);
        const assetPath = `${require('path').dirname(contentScriptPath)}/fence.css`;
        const fenceContent = 'strong { color: red; }';
        yield fs.writeFile(assetPath, fenceContent);
        const plugins = {
            'test': {
                id: 'test',
                contentScripts: {
                    [types_1.ContentScriptType.MarkdownItPlugin]: [
                        {
                            id: 'mdContentScript',
                            path: contentScriptPath,
                        },
                    ],
                },
                views: {},
            },
        };
        yield service.export({
            path: filePath,
            format: 'html',
            packIntoSingleFile: false,
            plugins,
        });
        const fenceRelativePath = 'pluginAssets/mdContentScript/fence.css';
        const content = yield fs.readFile(filePath, 'utf8');
        expect(content).toContain(fenceRelativePath);
        const readFenceContent = yield fs.readFile(`${(0, test_utils_1.exportDir)()}/${fenceRelativePath}`, 'utf8');
        expect(readFenceContent).toBe(fenceContent);
    })));
});
