// @joplin/lib/services/interop/InteropService_Exporter_Custom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const InteropService_Exporter_Base_1 = require("./InteropService_Exporter_Base");
class InteropService_Exporter_Custom extends InteropService_Exporter_Base_1.default {
    constructor(module) {
        super();
        this.module_ = null;
        this.module_ = module;
    }
    init(destPath, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.customContext_ = {
                destPath: destPath,
                options: options,
            };
            return this.module_.onInit(this.customContext_);
        });
    }
    processItem(itemType, item) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.module_.onProcessItem(this.customContext_, itemType, item);
        });
    }
    processResource(resource, filePath) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.module_.onProcessResource(this.customContext_, resource, filePath);
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.module_.onClose(this.customContext_);
        });
    }
}
exports.default = InteropService_Exporter_Custom;
