// @joplin/lib/commands/historyBackward.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runtime = exports.declaration = void 0;
const CommandService_1 = require("../services/CommandService");
const locale_1 = require("../locale");
exports.declaration = {
    name: 'historyBackward',
    label: () => (0, locale_1._)('Back'),
    iconName: 'icon-back',
};
const runtime = () => {
    return {
        execute: () => __awaiter(void 0, void 0, void 0, function* () {
            CommandService_1.utils.store.dispatch({
                type: 'HISTORY_BACKWARD',
            });
        }),
        enabledCondition: 'historyhasBackwardNotes',
    };
};
exports.runtime = runtime;
