// @joplin/lib/commands/openMasterPasswordDialog.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runtime = exports.declaration = void 0;
const locale_1 = require("../locale");
exports.declaration = {
    name: 'openMasterPasswordDialog',
    label: () => (0, locale_1._)('Manage master password...'),
    iconName: 'fas fa-key',
};
const runtime = () => {
    return {
        execute: (context, isOpen = true) => __awaiter(void 0, void 0, void 0, function* () {
            context.dispatch({
                type: 'DIALOG_OPEN',
                name: 'masterPassword',
                isOpen: isOpen,
            });
        }),
    };
};
exports.runtime = runtime;
