// @joplin/lib/services/plugins/ViewController.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class ViewController {
    constructor(handle, pluginId, store) {
        this.handle_ = handle;
        this.pluginId_ = pluginId;
        this.store_ = store;
    }
    get storeView() {
        return this.store_.getState().pluginService.plugins[this.pluginId_].views[this.handle];
    }
    get store() {
        return this.store_;
    }
    get pluginId() {
        return this.pluginId_;
    }
    get key() {
        return this.handle_;
    }
    get handle() {
        return this.handle_;
    }
    get type() {
        throw new Error('Must be overriden');
    }
    emitMessage(event) {
        return __awaiter(this, void 0, void 0, function* () {
            console.info('Calling ViewController.emitMessage - but not implemented', event);
        });
    }
    postMessage(message) {
        console.info('Calling ViewController.postMessage - but not implemented', message);
    }
}
exports.default = ViewController;
