// @joplin/lib/services/commands/stateToWhenClauseContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const reducer_1 = require("../../reducer");
const BaseModel_1 = require("../../BaseModel");
const Folder_1 = require("../../models/Folder");
const MarkupToHtml_1 = require("@joplin/renderer/MarkupToHtml");
const reducer_2 = require("../share/reducer");
function stateToWhenClauseContext(state, options = null) {
    options = Object.assign({ commandFolderId: '', commandNoteId: '' }, options);
    const selectedNoteId = state.selectedNoteIds.length === 1 ? state.selectedNoteIds[0] : null;
    const selectedNote = selectedNoteId ? BaseModel_1.default.byId(state.notes, selectedNoteId) : null;
    // const commandNoteId = options.commandNoteId || selectedNoteId;
    // const commandNote:NoteEntity = commandNoteId ? BaseModel.byId(state.notes, commandNoteId) : null;
    const commandFolderId = options.commandFolderId || state.selectedFolderId;
    const commandFolder = commandFolderId ? BaseModel_1.default.byId(state.folders, commandFolderId) : null;
    return {
        // Application state
        notesAreBeingSaved: reducer_1.stateUtils.hasNotesBeingSaved(state),
        syncStarted: state.syncStarted,
        // Current location
        inConflictFolder: state.selectedFolderId === Folder_1.default.conflictFolderId(),
        // Note selection
        oneNoteSelected: !!selectedNote,
        someNotesSelected: state.selectedNoteIds.length > 0,
        multipleNotesSelected: state.selectedNoteIds.length > 1,
        noNotesSelected: !state.selectedNoteIds.length,
        // Note history
        historyhasBackwardNotes: state.backwardHistoryNotes.length > 0,
        historyhasForwardNotes: state.forwardHistoryNotes.length > 0,
        // Folder selection
        oneFolderSelected: !!state.selectedFolderId,
        // Current note properties
        noteIsTodo: selectedNote ? !!selectedNote.is_todo : false,
        noteTodoCompleted: selectedNote ? !!selectedNote.todo_completed : false,
        noteIsMarkdown: selectedNote ? selectedNote.markup_language === MarkupToHtml_1.default.MARKUP_LANGUAGE_MARKDOWN : false,
        noteIsHtml: selectedNote ? selectedNote.markup_language === MarkupToHtml_1.default.MARKUP_LANGUAGE_HTML : false,
        // Current context folder
        folderIsShareRoot: commandFolder ? (0, reducer_2.isRootSharedFolder)(commandFolder) : false,
        folderIsShareRootAndNotOwnedByUser: commandFolder ? (0, reducer_2.isRootSharedFolder)(commandFolder) && !(0, reducer_2.isSharedFolderOwner)(state, commandFolder.id) : false,
        folderIsShareRootAndOwnedByUser: commandFolder ? (0, reducer_2.isRootSharedFolder)(commandFolder) && (0, reducer_2.isSharedFolderOwner)(state, commandFolder.id) : false,
        folderIsShared: commandFolder ? !!commandFolder.share_id : false,
        joplinServerConnected: [9, 10].includes(state.settings['sync.target']),
        hasMultiProfiles: state.profileConfig && state.profileConfig.profiles.length > 1,
    };
}
exports.default = stateToWhenClauseContext;
